jenkins/snapshot-elasticsearch.sh

```diff
@@ -0,0 +1,137 @@
+#!/bin/bash
+
+set -e
+
+# Show usage and fail if there is a problem with the command line.
+#
+function help {
+    local -r av0="$1" environment="$2" verb="$3"
+    local -r -a environments=(dev prod staging)
+    local -r -a tools=(curl date jq)
+    local -r -a verbs=(configure list take delete)
+    local -r -a usage=(
+        ''
+        "$av0: Snapshot an Elasticsearch cluster into GCS."
+        ''
+        "Usage: $av0 <environment> <verb> [<snapshot> ...]"
+        ''
+        "Where: <environment> is one of: ${environments[*]}"
+        "       <verb> is one of: ${verbs[*]}, and"
+        "       <snapshot> is a snapshot name from 'list'"
+        "       and where:"
+        "       configure configures snapshots in environment,"
+        "       list      lists the snapshots already taken, and"
+        "       take      takes a snapshot now, and"
+        "       delete    deletes the named snapshots."
+        ''
+        "Example: $av0 dev list"
+        ''
+        "BTW, you ran: $*"
+        ''
+    )
+    local ok=no x
+    for x in ${environments[*]}; do test $x = "$environment" && ok=yes; done
+    if test $ok = no
+    then
+        echo 1>&2 "${av0}: Unrecognized environment: '$environment'"
+    fi
+    ok=no
+    for x in ${verbs[*]}; do test $x = "$verb" && ok=yes; done
+    if test $ok = no
+    then
+        echo 1>&2 "${av0}: Unrecognized verb: '$verb'"
+    fi
+    for x in "${tools[@]}"; do which $x >/dev/null || ok=no; done
+    if test $ok = no
+    then
+        echo 1>&2 "${av0}: You need these tools on PATH: ${tools[*]}"
+    fi
+    if test $ok = no
+    then
+        for line in "${usage[@]}"; do echo "$line" 1>&2; done
+        exit 1
+    fi
+}
+
+# Report that $av0 is running the rest of arguments as a command line.
+# Then run that command line and pipe it through 'jq .'.
+#
+function echoRunJq () {
+    local -r av0="$1" ; shift
+    local -a command
+    local arg result
+    for arg in "$@"; do command+=("'$arg'"); done
+    echo $av0: Running "${command[@]}"
+    result=$("$@")
+    echo "$result" | jq .
+}
+
+# Configure snapshots to $bucket in for the $es cluster.
+#
+function snapshot_configure () {
+    local -r av0="$1" es=$2 bucket=$3
+    local -r data='
+    {
+        "type": "gcs",
+        "settings": {
+            "bucket": "'$bucket'",
+            "service_account": "_default_",
+            "compress": true,
+            "application_name": "clio-elasticsearch"
+        }
+    }'
+    echoRunJq "$av0" curl -s -X PUT --data "$data" $es/_snapshot/$bucket
+}
+
+# List the snapshots in $environment for $es cluster.
+#
+function snapshot_list () {
+    local -r av0="$1" es=$2 bucket=$3
+    echoRunJq "$av0" curl -s $es/_snapshot/$bucket/_all
+}
+
+# Take a snapshot in $environment for $es cluster.
+#
+function snapshot_take () {
+    local -r av0="$1" es=$2 bucket=$3
+    local -r jq='keys|.[]|select(startswith(".")|not)'
+    local -r snapshot=clio-$(date -u +%Y-%m-%d-%H-%M-%S)
```

Comment:
Our usage is clio specific but in theory this could be used to backup any ES server, correct?

Reply:
Probably.
I don't want to generalize too soon, but could name snapshots just 'snap-...'.
(I'd prefix the snapshot names with the bucket name if it were shorter.

Reply:
OK. Now I've generalized too soon. ;-)